util/topsql/reporter/reporter.go

```diff
@@ -0,0 +1,411 @@
+	// NOTE: if use multiple worker goroutine, we should make sure that access to the digest map is thread-safe
+	for {
+		job := <-tsc.planRegisterChan
+		tsc.mu.RLock()
+		_, exist := tsc.normalizedPlanMap[job.planDigest]
+		tsc.mu.RUnlock()
+		if exist {
+			continue
+		}
+		planDecoded, err := tsc.planBinaryDecoder(job.normalizedPlan)
+		if err != nil {
+			fmt.Printf("decode plan failed: %v\n", err)
+			continue
+		}
+		tsc.normalizedPlanMap[job.planDigest] = planDecoded
+	}
+}
+
+// snapshot will collect the current snapshot of data for transmission
+// This could run in parallel with `Collect()`, so we should guard it by a mutex.
+//
+// NOTE: we could optimize this using a mutex protected pointer-swapping operation,
+// which means we maintain 2 *struct which contains the maps, and snapshot() atomically
+// swaps the pointer. After this, the writing is shifted to the new struct, and
+// we can do the snapshot in the background.
+func (tsc *TopSQLReporterImpl) snapshot() []*tipb.CollectCPUTimeRequest {
+	tsc.mu.RLock()
+	defer tsc.mu.RUnlock()
+	total := len(tsc.topSQLMap)
+	batch := make([]*tipb.CollectCPUTimeRequest, total)
+	i := 0
+	for _, value := range tsc.topSQLMap {
+		normalizedSQL := tsc.normalizedSQLMap[string(value.SQLDigest)]
+		normalizedPlan := tsc.normalizedPlanMap[string(value.PlanDigest)]
+		batch[i] = &tipb.CollectCPUTimeRequest{
+			TimestampList:  value.TimestampList,
+			CpuTimeMsList:  value.CPUTimeMsList,
+			SqlDigest:      value.SQLDigest,
+			NormalizedSql:  normalizedSQL,
+			PlanDigest:     value.PlanDigest,
+			NormalizedPlan: normalizedPlan,
+		}
+		i++
+	}
+	return batch
+}
+
+// sendBatch sends a batch of TopSQL records streamingly.
+// TODO: benchmark test with large amount of data (e.g. 5000), tune with grpc.WithWriteBufferSize()
+func (tsc *TopSQLReporterImpl) sendBatch(stream tipb.TopSQLAgent_CollectCPUTimeClient, batch []*tipb.CollectCPUTimeRequest) error {
+	for _, req := range batch {
+		if err := stream.Send(req); err != nil {
+			log.Error("TopSQL: send stream request failed, %v", zap.Error(err))
+			return err
+		}
+	}
+	// response is Empty, drop it for now
+	_, err := stream.CloseAndRecv()
+	if err != nil {
+		log.Error("TopSQL: receive stream response failed, %v", zap.Error(err))
+		return err
+	}
+	return nil
+}
+
+// sendToAgentWorker will send a snapshot to the gRPC endpoint every collect interval
+func (tsc *TopSQLReporterImpl) sendToAgentWorker() {
+	ticker := time.NewTicker(tsc.collectInterval)
+	for {
+		var ctx context.Context
+		var cancel context.CancelFunc
+		select {
+		case <-ticker.C:
+			batch := tsc.snapshot()
+			// NOTE: Currently we are creating/destroying a TCP connection to the agent every time.
+			// It's fine if we do this every minute, but need optimization if we need to do it more frequently, like every second.
+			conn, client, err := newAgentClient(tsc.agentGRPCAddress)
+			if err != nil {
+				log.Error("TopSQL: failed to create agent client, %v", zap.Error(err))
```

Comment:
zap doesn't work in this way. It should be:

```suggestion
				log.Warn("TopSQL: failed to create agent client", zap.Error(err))
```

Reply:
fixed